src/sqlfluff/core/rules/surrogates/segments.py

```diff
@@ -0,0 +1,93 @@
+"""Surrogate class for working with Segment collections."""
+from typing import Callable, List, Optional, Sequence, Type, TypeVar
+
+from sqlfluff.core.parser import BaseSegment
+from sqlfluff.core.rules.base import LintFix
+from sqlfluff.core.templaters.base import TemplatedFile
+from sqlfluff.core.rules.surrogates.raw_file_slice import RawFileSlices
+
+Predicate = TypeVar("Predicate", str, Type, Callable[[BaseSegment], bool])
+
+
+class Segments:
+    """Encapsulates a sequence of one or more BaseSegments.
+
+    The segments may or may not be contiguous in a parse tree.
+    Provides useful operations on a sequence of segments to simplify rule creation.
+    """
+
+    def __init__(self, templated_file: Optional[TemplatedFile], *segments: BaseSegment):
+        self.templated_file = templated_file
+        self.segments = segments
+
+    def all(self, *predicates: Predicate) -> bool:
+        """Do all the segments match?"""
+        cp = _CompositePredicate(*predicates)
+        return all(cp(s) for s in self.segments)
+
+    def any(self, *predicates: Predicate) -> bool:  # pragma: no cover
+        """Do any of the segments match?"""
+        cp = _CompositePredicate(*predicates)
+        return any(cp(s) for s in self.segments)
+
+    @property
+    def raw_slices(self) -> RawFileSlices:
+        """Raw slices of the segments."""
+        if not self.templated_file:  # pragma: no cover
+            raise ValueError(
+                'Segments.raw_slices: "templated_file" property is required.'
+            )
+        raw_slices = set()
+        for s in self.segments:
+            source_slice = s.pos_marker.source_slice
+            raw_slices.update(
+                self.templated_file.raw_slices_spanning_source_slice(source_slice)
+            )
+        return RawFileSlices(self.templated_file, *raw_slices)
+
+    # def with_children(self) -> "Segments":
+    #     """Returns an object that includes first-level children."""
+    #     raise NotImplementedError
+
+    # def with_descendants(self) -> "Segments":
+    #     """Returns an object that includes all descendants."""
+    #     raise NotImplementedError
+
+    # def select(
+    #     self,
+    #     start_seg: Optional[BaseSegment] = None,
+    #     stop_seg: Optional[BaseSegment] = None,
+    #     select_if=Optional[Sequence[Predicate]],
+    #     loop_while=Optional[Sequence[Predicate]],
+    # ) -> "Segments":
+    #     """Retrieve range/subset."""
+    #     raise NotImplementedError
+
+    def delete(self, *predicates: Predicate) -> Sequence["LintFix"]:
+        """Return LintFix objects to delete segments that satisfy predicates."""
+        cp = _CompositePredicate(*predicates)
+        return [LintFix("delete", s) for s in self.segments if cp(s)]
```

Comment:
Would we also have `create_before`, `create_after`, and `edit` methods? They all anchor to a single segment and add something new so don't really make as much sense if there are multiple segments in this class?

Curious to see what the plan is here as it would be good to be consistent across the different LintFix types. If we didn't add those as methods to this class then this method name should probably be more explicit like `get_delete_fixes` so that it's clear in the rule code that this just wraps the delete fix type when we have mixes of the different types.

i.e.
```
edit_fix = LintFix("edit", <some_segment>, [<edit_segments>,...])
delete_fix = segments.get_delete_fixes()
```
is more clear when mixing

Reply:
I haven't thought that far ahead re: the other edit types. Was just experimenting to see what it "felt? like to have a method create the "delete" `LintFix` objects rather than having to explicitly create it yourself. I think it's an improvement, so I hope the other ones are a natural fit as well. I think **one** of the usability issues in the current way of doing things is that you have to create and work with various unrelated objects. It feels disjointed, verbose,  and it disrupts your "flow" when writing rules.

If we can connect these together so you're mostly working with one object plus others that it returns, it's a simpler, more unified experience.

Reply:
Thinking more about this. As you say, the other edit types only apply to a single segment. Therefore, they don't make sense as instance methods. We could add them as class or static methods for completeness. Think that's worth doing now? Or wait and think about it more?

Reply:
tbh I feel like they don't belong in that class. To me segments feels like it should just be for filtering and querying lists of segments.

So maybe we should just let LintFix take in Segments type?
e.g.
```python
raw_stack = context.surrogates.raw_stack
LintFix("delete", s.select(select_if=[cp.is_whitespace], loop_while=[cp.not(cp.is_whitespace)])
```
allows you to filter segments easily and then LintFix delete is still consistent with the other methods.

Reply:
I think an easy and useful way to do this is to make `Segments` iterable and return BaseSegments.

So
```python
[s for s in Segments(<input_segments>).select(<some_filter_condition>)]
```
easily gets you back to a list of Base/Raw segments
or more concisely
```python
list(Segments(<input_segments>).select(<some_filter_condition>)): List[BaseSegment]
```

and you can just type LintFix input as `Iterable[BaseSegment]` to avoid worrying about manually converting back to segments in rule code.

Reply:
Basically keeping it a filter language. It can still have logic to add in extra segments etc but just not build the LintfFixes

Reply:
related LintFix improvement in #2133